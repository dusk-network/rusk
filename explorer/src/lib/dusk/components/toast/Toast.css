.dusk-toast {
  display: flex;
  flex-direction: column;
  position: fixed;
  right: 20px;
  bottom: 20px;
  z-index: 5;
  row-gap: 0.625em;
}

.dusk-toast__item {
  display: flex;
  column-gap: 0.625em;
  padding: 0.625em;
  border-radius: 1em;
  background-color: var(--background-color-alt);
  min-width: 200px;
  max-width: 250px;
}

.dusk-toast__item-icon-wrapper {
  display: flex;
  align-items: center;
  justify-content: center;
  width: 1.625em;
  height: 1.625em;
  border-radius: 100%;
  flex-shrink: 0;
}

.dusk-toast__item-icon-wrapper--info {
  background-color: var(--info-color);
}

.dusk-toast__item-icon-wrapper--success {
  background-color: var(--success-color);
}

.dusk-toast__item-icon-wrapper--warning {
  background-color: var(--warning-color);
}

.dusk-toast__item-icon-wrapper--error {
  background-color: var(--error-color);
}

.dusk-toast__item-icon-wrapper--error .dusk-toast__item-icon {
  color: #fff;
}
