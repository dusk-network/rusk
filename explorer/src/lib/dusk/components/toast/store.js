import { get, writable } from "svelte/store";

/**
 * @type {import('svelte/store').Writable<ToastItem[]>}
 * @description Stores each toast as an object in the array
 */
export const toastList = writable([]);

/**
 * @type {import('svelte/store').Writable<Number>}
 * @description Stores the timer fused by all toasts
 */
export const toastTimer = writable(0);

/**
 * @param {string} message
 * @param {string} icon
 * @param {TooltipType} type
 */

function addToast(type, message, icon) {
  const id = `dusk-toast-${Math.random().toString(36)}`;

  toastList.update((store) => {
    return [
      ...store,
      {
        icon: icon,
        id: id,
        message: message,
        type: type,
      },
    ];
  });

  const timeoutID = window.setTimeout(() => {
    deleteToast(id, timeoutID);
  }, get(toastTimer));
}

/**
 * @param {string} id
 * @param {number} timeout
 */
function deleteToast(id, timeout) {
  window.clearTimeout(timeout);

  // Deletes toast from store queue
  toastList.update((store) => {
    return store.filter((toast) => toast.id !== id);
  });
}

export const toast = addToast;
