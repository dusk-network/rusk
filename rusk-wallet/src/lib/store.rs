// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.
//
// Copyright (c) DUSK NETWORK. All rights reserved.

use std::io::Read;
use std::{fmt, fs};
use std::path::PathBuf;
use bytes::{BytesMut, BufMut, Bytes, Buf};

use dusk_pki::{SecretSpendKey, SecretKey};
use dusk_wallet_core::{Store, derive_sk, derive_ssk};

use crate::lib::errors::CliError;

/// Default directory name to store settings and keystore
const DATA_DIR: &str = ".dusk";

/// Stores all the user's settings and keystore in the file system
pub struct LocalStore {
    name: String,
    path: PathBuf,
    seed: [u8; 64],
}

impl Store for LocalStore {

    type Error = CliError;

    /// Retrieves the seed used to derive keys.
    fn get_seed(&self) -> Result<[u8; 64], Self::Error> {
        Ok(self.seed)
    }

    /// Retrieves a derived secret spend key from the store.
    fn retrieve_ssk(&self, index: u64) -> Result<SecretSpendKey, Self::Error> {
        Ok(derive_ssk(&self.seed, index))
    }

    /// Retrieves a derived secret key from the store.
    fn retrieve_sk(&self, index: u64) -> Result<SecretKey, Self::Error> {
        Ok(derive_sk(&self.seed, index))
    }

}

impl LocalStore {

    /// Creates a new store
    pub fn new(name: String, seed: [u8;64]) -> Result<LocalStore, CliError> {

        // construct path to file
        let home = dirs::home_dir().unwrap();
        let mut path = PathBuf::new();
        path.push(home.as_os_str());
        path.push(DATA_DIR);
        path.push(&name);
        path = path.with_extension("dat");

        // get the seed
        let mut seed_bytes= [0u8; 64];
        seed_bytes.copy_from_slice(&seed);

        // create the local store
        let store = LocalStore{
            name: name,
            path: path,
            seed: seed_bytes,
        };

        // save to disk and return
        store.save()?;
        Ok(store)

    }


    /// Loads wallet file from file
    pub fn from_file(path: PathBuf, _pwd: String) -> Result<LocalStore, CliError> {

        // basic sanity check
        if !path.is_file() {
            return Err(CliError::CorruptedFile);
        }

        // attempt to load and decode wallet
        let data = Bytes::from(fs::read(&path).unwrap());
        let mut reader = data.reader();

        // wallet_seed
        let mut wallet_seed = [0u8; 64];
        reader.read(&mut wallet_seed)?;

        // extract the name
        let name = path.file_name()
            .ok_or(CliError::CorruptedFile)?.to_str()
            .ok_or(CliError::CorruptedFile)?;

        // create and return
        Ok(LocalStore{
            name: String::from(name),
            path: path,
            seed: wallet_seed,
        })

    }

    /// Saves wallet to a file
    pub fn save(&self) -> Result<(), CliError> {

        let mut buf = BytesMut::new();

        // wallet seed (todo: encrypt)
        buf.put_slice(&self.seed);

        // write file
        fs::write(&self.path, &buf[..]).unwrap();
        Ok(())

    }

    /// The name of the wallet
    pub fn name(&self) -> String {
        self.name.clone()
    }

}

impl fmt::Debug for LocalStore {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "-------------------\n\
            LocalStore: {}\n\
            Name: {}\n\
            Seed: {:?}",
            self.path.as_os_str().to_str().unwrap(), 
            self.name, self.seed)
    }
}