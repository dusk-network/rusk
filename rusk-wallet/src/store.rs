// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.
//
// Copyright (c) DUSK NETWORK. All rights reserved.

use wallet_core::Seed;

use crate::clients::State;

/// Provides a valid wallet seed to `wallet_core`
#[derive(Clone)]
pub(crate) struct LocalStore {
    seed: Seed,
}

impl LocalStore {
    /// Retrieves the seed used to derive keys.
    pub fn get_seed(&self) -> &Seed {
        &self.seed
    }

    pub fn inner_mut(&mut self) -> &mut Seed {
        &mut self.seed
    }
}

impl From<Seed> for LocalStore {
    fn from(seed: Seed) -> Self {
        LocalStore { seed }
    }
}

impl State {
    /// Retrieves the seed used to derive keys.
    pub fn get_seed(&self) -> &Seed {
        self.store().get_seed()
    }
}
