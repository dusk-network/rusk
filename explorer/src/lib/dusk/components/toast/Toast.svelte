<script>
  import { flip } from "svelte/animate";
  import { fly } from "svelte/transition";
  import { Icon } from "$lib/dusk/components";
  import { makeClassName } from "$lib/dusk/string";
  import { toastList, toastTimer } from "./store";
  import { onMount } from "svelte";

  import "./Toast.css";

  /** @type {String | Undefined} */
  export let className = undefined;

  /** @type {Number} */
  export let timer = 2000;

  /** @type {Number} */
  export let flyDuration = 500;

  const classes = makeClassName(["dusk-toast", className]);

  onMount(() => {
    $toastTimer = timer;
  });
</script>

<ul {...$$restProps} class={classes}>
  {#each $toastList as { id, icon, message, type } (id)}
    <li
      in:fly|global={{ duration: flyDuration, x: 200 }}
      out:fly|global={{ duration: flyDuration, x: 200 }}
      animate:flip={{ duration: 200 }}
      class="dusk-toast__item"
    >
      {#if icon}
        <span
          class={`dusk-toast__item-icon-wrapper dusk-toast__item-icon-wrapper--${type}`}
        >
          <Icon className="dusk-toast__item-icon" path={icon} size="default" />
        </span>
      {/if}
      <span class="dusk-toast__item-message">
        {message}
      </span>
    </li>
  {/each}
</ul>
