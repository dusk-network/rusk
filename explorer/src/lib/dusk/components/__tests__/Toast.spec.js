import { afterAll, afterEach, describe, expect, it, vi } from "vitest";
import { cleanup, render } from "@testing-library/svelte";
import { get } from "svelte/store";
import { fly } from "svelte/transition";
import { mdiAlertOutline } from "@mdi/js";

import { Toast } from "..";
import { toast, toastList } from "../toast/store";

vi.mock("svelte/transition");

describe("Toast", () => {
  const baseProps = {
    flyDuration: 500,
    timer: 2000,
  };

  /** @param {HTMLElement} list */
  const getToastItems = (list) => list.querySelectorAll(".dusk-toast__item");

  vi.useFakeTimers();

  afterEach(() => {
    vi.mocked(fly).mockClear();
    cleanup();
  });

  afterAll(() => {
    vi.doUnmock("svelte/transition");
  });

  it("should render and dismiss the Toast component with the correct properties", async () => {
    const { getByRole } = render(Toast, baseProps);
    const list = getByRole("list");

    expect(get(toastList).length).toBe(0);
    expect(getToastItems(list).length).toBe(0);
    expect(fly).not.toHaveBeenCalled();

    toast("success", "Render Toast 1", mdiAlertOutline);

    await vi.advanceTimersToNextTimerAsync();

    const items = getToastItems(list);
    const toastStoredList = get(toastList);

    expect(toastStoredList.length).toBe(1);
    expect(toastStoredList[0]).toStrictEqual({
      icon: mdiAlertOutline,
      id: expect.any(String),
      message: "Render Toast 1",
      type: "success",
    });
    expect(items.length).toBe(1);
    expect(items[0]).toHaveTextContent("Render Toast 1");
    expect(
      items[0].querySelector(".dusk-toast__item-icon-wrapper--success")
    ).toBeDefined();
    expect(
      items[0].querySelector(`path[d="${mdiAlertOutline}"]`)
    ).toBeDefined();
    expect(fly).toHaveBeenCalledTimes(1);
    expect(fly).toHaveBeenCalledWith(
      items[0],
      expect.objectContaining({ duration: baseProps.flyDuration }),
      { direction: "in" }
    );
    expect(list).toMatchSnapshot();

    await vi.advanceTimersByTimeAsync(baseProps.timer);

    expect(get(toastList).length).toBe(0);
    expect(fly).toHaveBeenCalledTimes(2);
    expect(fly).toHaveBeenNthCalledWith(
      2,
      items[0],
      expect.objectContaining({ duration: baseProps.flyDuration }),
      { direction: "out" }
    );
  });

  it("should render the Toast component with the correct custom class and rest props", async () => {
    const customProps = {
      className: "test-toast",
      "data-baz": "baz",
      flyDuration: 500,
      id: "toast-container",
      timer: 2000,
    };

    const { getByRole } = render(Toast, customProps);
    const list = getByRole("list");

    expect(list).toHaveClass(`${customProps.className}`);
    expect(list).toHaveAttribute("id", customProps.id);
    expect(list).toHaveAttribute("data-baz", customProps["data-baz"]);

    expect(list).toMatchSnapshot();
  });

  it("should render and dismiss 2 Toasts", async () => {
    const { getByRole } = render(Toast, baseProps);
    const list = getByRole("list");

    expect(get(toastList).length).toBe(0);
    expect(getToastItems(list).length).toBe(0);
    expect(fly).not.toHaveBeenCalled();

    toast("success", "Render Toast 1", mdiAlertOutline);
    toast("info", "Render Toast 2", mdiAlertOutline);

    await vi.advanceTimersToNextTimerAsync();

    const items = getToastItems(list);
    const toastStoredList = get(toastList);

    expect(toastStoredList.length).toBe(2);
    expect(toastStoredList[0]).toStrictEqual({
      icon: mdiAlertOutline,
      id: expect.any(String),
      message: "Render Toast 1",
      type: "success",
    });
    expect(toastStoredList[1]).toStrictEqual({
      icon: mdiAlertOutline,
      id: expect.any(String),
      message: "Render Toast 2",
      type: "info",
    });
    expect(items.length).toBe(2);
    expect(items[0]).toHaveTextContent("Render Toast 1");
    expect(
      items[0].querySelector(".dusk-toast__item-icon-wrapper--success")
    ).toBeDefined();
    expect(
      items[0].querySelector(`path[d="${mdiAlertOutline}"]`)
    ).toBeDefined();

    expect(items[1]).toHaveTextContent("Render Toast 2");
    expect(
      items[1].querySelector(".dusk-toast__item-icon-wrapper--info")
    ).toBeDefined();
    expect(
      items[1].querySelector(`path[d="${mdiAlertOutline}"]`)
    ).toBeDefined();

    expect(fly).toHaveBeenCalledTimes(2);
    expect(fly).toHaveBeenCalledWith(
      items[0],
      expect.objectContaining({ duration: baseProps.flyDuration }),
      { direction: "in" }
    );
    expect(fly).toHaveBeenCalledWith(
      items[1],
      expect.objectContaining({ duration: baseProps.flyDuration }),
      { direction: "in" }
    );
    expect(list).toMatchSnapshot();

    await vi.advanceTimersByTimeAsync(baseProps.timer);

    expect(get(toastList).length).toBe(0);
    expect(fly).toHaveBeenCalledTimes(4);
    expect(fly).toHaveBeenNthCalledWith(
      3,
      items[0],
      expect.objectContaining({ duration: baseProps.flyDuration }),
      { direction: "out" }
    );
    expect(fly).toHaveBeenNthCalledWith(
      4,
      items[1],
      expect.objectContaining({ duration: baseProps.flyDuration }),
      { direction: "out" }
    );
  });
});
