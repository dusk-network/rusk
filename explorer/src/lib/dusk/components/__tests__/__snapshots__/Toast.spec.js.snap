// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Toast > should render and dismiss 2 Toasts 1`] = `
<ul
  class="dusk-toast"
>
  <li
    class="dusk-toast__item"
  >
    <span
      class="dusk-toast__item-icon-wrapper dusk-toast__item-icon-wrapper--success"
    >
      <svg
        class="dusk-icon dusk-icon--size--default dusk-toast__item-icon"
        role="graphics-symbol"
        viewBox="0 0 24 24"
      >
        <path
          d="M12,2L1,21H23M12,6L19.53,19H4.47M11,10V14H13V10M11,16V18H13V16"
        />
      </svg>
      
    </span>
     
    <span
      class="dusk-toast__item-message"
    >
      Render Toast 1
    </span>
     
  </li>
  <li
    class="dusk-toast__item"
  >
    <span
      class="dusk-toast__item-icon-wrapper dusk-toast__item-icon-wrapper--info"
    >
      <svg
        class="dusk-icon dusk-icon--size--default dusk-toast__item-icon"
        role="graphics-symbol"
        viewBox="0 0 24 24"
      >
        <path
          d="M12,2L1,21H23M12,6L19.53,19H4.47M11,10V14H13V10M11,16V18H13V16"
        />
      </svg>
      
    </span>
     
    <span
      class="dusk-toast__item-message"
    >
      Render Toast 2
    </span>
     
  </li>
</ul>
`;

exports[`Toast > should render and dismiss the Toast component with the correct properties 1`] = `
<ul
  class="dusk-toast"
>
  <li
    class="dusk-toast__item"
  >
    <span
      class="dusk-toast__item-icon-wrapper dusk-toast__item-icon-wrapper--success"
    >
      <svg
        class="dusk-icon dusk-icon--size--default dusk-toast__item-icon"
        role="graphics-symbol"
        viewBox="0 0 24 24"
      >
        <path
          d="M12,2L1,21H23M12,6L19.53,19H4.47M11,10V14H13V10M11,16V18H13V16"
        />
      </svg>
      
    </span>
     
    <span
      class="dusk-toast__item-message"
    >
      Render Toast 1
    </span>
     
  </li>
</ul>
`;

exports[`Toast > should render the Toast component with the correct custom class and rest props 1`] = `
<ul
  class="dusk-toast test-toast"
  data-baz="baz"
  id="toast-container"
/>
`;
